File: packages/reliai-node/src/pipeline.ts
import { getDefaultClient, type ReliaiSpanScope } from "./client.ts";

export interface ReliaiPipelineMetadata {
  spanType?: string;
  model?: string;
  tokens?: number;
  vector_db?: string;
  tool_name?: string;
  prompt_template?: string;
  [key: string]: unknown;
}

async function runPipelineSpan<T>(
  name: string,
  spanType: string,
  fn: (scope: ReliaiSpanScope) => Promise<T> | T,
  metadata: ReliaiPipelineMetadata = {},
): Promise<T> {
  return getDefaultClient().span(
    name,
    fn,
    {
      ...metadata,
      span_type: metadata.spanType ?? spanType,
    },
  );
}

export async function retrieval<T>(
  fn: (scope: ReliaiSpanScope) => Promise<T> | T,
  metadata: ReliaiPipelineMetadata = {},
): Promise<T> {
  return runPipelineSpan("retrieval", "retrieval", fn, metadata);
}

export async function promptBuild<T>(
  fn: (scope: ReliaiSpanScope) => Promise<T> | T,
  metadata: ReliaiPipelineMetadata = {},
): Promise<T> {
  return runPipelineSpan("prompt_build", "prompt", fn, metadata);
}

export async function llmCall<T>(
  fn: (scope: ReliaiSpanScope) => Promise<T> | T,
  metadata: ReliaiPipelineMetadata = {},
): Promise<T> {
  return runPipelineSpan("llm_call", "llm", fn, metadata);
}

export async function toolCall<T>(
  fn: (scope: ReliaiSpanScope) => Promise<T> | T,
  metadata: ReliaiPipelineMetadata = {},
): Promise<T> {
  return runPipelineSpan("tool_call", "tool", fn, metadata);
}

export async function postprocess<T>(
  fn: (scope: ReliaiSpanScope) => Promise<T> | T,
  metadata: ReliaiPipelineMetadata = {},
): Promise<T> {
  return runPipelineSpan("postprocess", "postprocess", fn, metadata);
}

File: packages/reliai-python/reliai/pipeline.py
from __future__ import annotations

from collections.abc import Callable
from typing import Any, TypeVar

from .client import get_default_client

T = TypeVar("T")


def _run_pipeline_step(name: str, span_type: str, fn: Callable[[], T], metadata: dict[str, Any] | None = None) -> T:
    with get_default_client().span(name, {"span_type": span_type, **(metadata or {})}):
        return fn()


def retrieval(fn: Callable[[], T], metadata: dict[str, Any] | None = None) -> T:
    return _run_pipeline_step("retrieval", "retrieval", fn, metadata)


def prompt_build(fn: Callable[[], T], metadata: dict[str, Any] | None = None) -> T:
    return _run_pipeline_step("prompt_build", "prompt", fn, metadata)


def llm_call(fn: Callable[[], T], metadata: dict[str, Any] | None = None) -> T:
    return _run_pipeline_step("llm_call", "llm", fn, metadata)


def tool_call(fn: Callable[[], T], metadata: dict[str, Any] | None = None) -> T:
    return _run_pipeline_step("tool_call", "tool", fn, metadata)


def postprocess(fn: Callable[[], T], metadata: dict[str, Any] | None = None) -> T:
    return _run_pipeline_step("postprocess", "postprocess", fn, metadata)
